File: 241-different-ways-to-add-parentheses/different-ways-to-add-parentheses.py
class Solution:
    def diffWaysToCompute(self, expression: str) -> List[int]:
        @lru_cache(None)
        def ways(expr):
            results = []
            for i, c in enumerate(expr):
                if c in "+-*":
                    left = ways(expr[:i])
                    right = ways(expr[i+1:])
                    for l in left:
                        for r in right:
                            if c == '+':
                                results.append(l + r)
                            elif c == '-':
                                results.append(l - r)
                            else:  # '*'
                                results.append(l * r)
            if not results:  # base case: no operator, just a number
                results.append(int(expr))
            return results

        return ways(expression)

File: 241-different-ways-to-add-parentheses/different-ways-to-add-parentheses.java
import java.util.*;

class Solution {
    private Map<String, List<Integer>> memo = new HashMap<>();
    
    public List<Integer> diffWaysToCompute(String expression) {
        if (memo.containsKey(expression)) {
            return memo.get(expression);
        }
        
        List<Integer> results = new ArrayList<>();
        
        for (int i = 0; i < expression.length(); i++) {
            char c = expression.charAt(i);
            if (c == '+' || c == '-' || c == '*') {
                List<Integer> left = diffWaysToCompute(expression.substring(0, i));
                List<Integer> right = diffWaysToCompute(expression.substring(i + 1));
                
                for (int l : left) {
                    for (int r : right) {
                        if (c == '+') {
                            results.add(l + r);
                        } else if (c == '-') {
                            results.add(l - r);
                        } else { // '*'
                            results.add(l * r);
                        }
                    }
                }
            }
        }
        
        // base case: no operator, just a number
        if (results.isEmpty()) {
            results.add(Integer.parseInt(expression));
        }
        
        memo.put(expression, results);
        return results;
    }
}

